AY2425S1-CS2103T-W09-3/tp
# Add sort by date feature

## Description

Closes [Add sort by date feature #107](https://github.com/AY2425S1-CS2103T-W09-3/tp/issues/107)
Closes [Date field should not be in the future #111](https://github.com/AY2425S1-CS2103T-W09-3/tp/issues/111)


This PR introduces several updates and improvements to the project by adding SortCommand, SortCommandParser, DateComparator, SortCommandParserTest, DateTest, DateComparatorTest.


These changes ensure more robust functionality for handling users' internship applications list as they will be able to sort their list by earliest or latest applications first through the use of the /sort command

## Changes made

- SortCommand: Added the command to handle the sorting of the internship application list.
- SortCommandTest: Added tests to verify the correct behavior of the SortCommand.
- SortCommandParser: Parses input arguments and creates the appropriate SortCommand.
- SortCommandParserTest: Added tests to ensure the SortCommandParser correctly parses valid inputs and handles errors for invalid inputs.
- DateComparator: Added this class to handle the sorting order.
- DateComparatorTest: Added tests to ensure the DateComparator correctly enables the sorting order that the user may want.
- DateTest: Added tests to ensure the Date class functions correctly and validates the input correctly.

**Before:**
No support for handling sorting in HireMe.
Users are able to add an internship application with a future date.

**After:**
Full support for sorting by date in the system.
Test cases added for sort command parsing and execution, ensuring improved code coverage and reliability.
Users can only add internship applications with present or past dates.

## Impact of changes


- Increased Code Coverage: More test cases ensure that date, date comparator and sort command are thoroughly tested.
- Fixed bug where users are able to add an internship application with a future date.
- Code Robustness: Added test cases help catch potential edge cases and errors, making the codebase more reliable.

## Testing Steps


1. Run the full test suite (SortCommandTest, SortCommandParserTest, DateComparatorTest, DateTest) to ensure all tests pass and the system handles the sorting commands as expected.
2. Test manually by:
      - Navigate to the application's command interface.
      - Execute /sort earliest, /sort latest commands with various internship applications.
      - Verify that the ordering of the internship applications list is updated correctly.
      - Test with invalid inputs like /sort a, /sort, and ensure that appropriate error messages are displayed.
